### PR TITLE
Fixes for using Python APIs from Rust.

### PR DESCRIPTION
**Pull request by jroesch**
cc @hypercubestart this is a bigger PR to fix the current bindings and enable calling them with the Python code, but it needs a bit of cleanup, there are some changes that I used to debug segfaults.

### CODE REVIEW

**Comment by jroesch** [comment]
cc @hypercubestart @binarybana @imalsogreg updated this

File: python/tvm/relay/analysis/__init__.py [hunk 1]
```diff
@@ -26,7 +26,7 @@
 from . import call_graph
 from .call_graph import CallGraph
 
-# Feature
+# # Feature
```
Comment (imalsogreg) [inline python/tvm/relay/analysis/__init__.py:29]: Two `#`'s?

File: python/tvm/relay/build_module.py [hunk 2]
```diff
@@ -193,6 +193,18 @@ def get_params(self):
             ret[key] = value.data
         return ret
 
+@register_func("tvm.relay.build")
+def _rust_build_module(mod, target=None, target_host=None, params=None, mod_name="default"):
+    print(mod)
+    print("\n")
+    rt_mod = build(mod, target, target_host, params, mod_name).module
+    print(rt_mod)
+    print(rt_mod["default"])
```
Comment (imalsogreg) [inline python/tvm/relay/build_module.py:202]: leftover prints, or something we want to keep?

File: rust/tvm/README.md [hunk 3]
```diff
@@ -15,221 +15,40 @@
...
-    // get the set_input function
-    let ref set_input_fn = graph_runtime_module
-        .get_function("set_input", false)
-        .unwrap();
+The Rust bindings are composed of a few crates:
+- The [tvm](https://tvm.apache.org/docs/api/rust/tvm/index.html) crate which exposes Rust bindings to
+  both the compiler and runtime.
+- The [tvm_macros](https://tvm.apache.org/docs/api/rust/tvm/index.html) crate which provides macros
+  which generate unsafe boilerplate for TVM's data structures.
+- The [tvm_rt](https://tvm.apache.org/docs/api/rust/tvm_rt/index.html) crate which exposes Rust
```
Comment (imalsogreg) [inline rust/tvm/README.md:37]: Just wondering as a naive reader, why there is a `tvm_rt` crate when `tvm` exposes bindings to the runtime. (Is there _extra_ runtime stuff in `tmv_rt`, vs. `tvm_rt` being a subset of `tvm` for users that don't want to pull in the compiler bindings?

Comment (jroesch) [reply rust/tvm/README.md:37]: TVM generates two dynamic libraries when we build it, the runtime bindings can be used to link to just the TVM runtime bindings, while the full tvm package links both the runtime and the compiler.

File: python/tvm/relay/build_module.py (hunk 2 above)
Comment (jroesch) [reply python/tvm/relay/build_module.py:202]: leftover, going to try and land this now finally

**Review by hypercubestart — COMMENTED** [review]
does this test pass now?

File: rust/tvm/src/python.rs [hunk 4]
```diff
@@ -29,19 +29,30 @@ use pyo3::prelude::*;
...
+fn import_python<'p, 'b: 'p>(py: Python<'p>, to_import: &'b str) -> PyResult<&'p PyModule> {
+    let imported_mod = py.import(to_import)?;
+    Ok(imported_mod)
+}
 
 fn load_python_tvm_(py: Python) -> PyResult<String> {
-    let sys = py.import("tvm")?;
-    let version: String = sys.get("__version__")?.extract()?;
-    // py.run(TVMC_CODE, None, None)?;
+    let imported_mod = import_python(py, "tvm")?;
```
Comment (hypercubestart) [inline rust/tvm/src/python.rs:54]: oops forgot to add the comment. 

theres an ignored test `test_run` in this file, if everything works the ignore attribute should be removed

Comment (jroesch) [reply rust/tvm/src/python.rs:54]: This test seems to pass on most platforms but not CI, going to keep disabled until I can be sure it won't cause CI flakiness.

File: rust/tvm/src/ir/expr.rs [hunk 5]
```diff
@@ -32,12 +32,14 @@ use super::span::Span;
 #[type_key = "Expr"]
```
Comment (MarisaKirisame) [inline rust/tvm/src/ir/expr.rs:32]: would it make sense to make a macro for the boilerplate?

Comment (jroesch) [reply rust/tvm/src/ir/expr.rs:32]: This is a macro 🤯

Comment (MarisaKirisame) [reply rust/tvm/src/ir/expr.rs:32]: I see. I just see the span boilerplate and thought you can refactor them. cool in this case then.

**Comment by vegaluisjose** [comment]
There seems to be a problem with scipy in the CI @jroesch ?

**Review by vegaluisjose — APPROVED** [review]
LGTM

File: rust/tvm/src/python.rs (hunk 4 above)
Comment (hypercubestart) [reply rust/tvm/src/python.rs:54]: ok, its strange to me that loading tvm into rust is flaky though

File: rust/tvm/src/compiler/graph_rt.rs [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    }
+}
+
+/// Compile a module from a configuration and IRModule.
+///
+/// # Arguments
+///
+/// * `config` - The configuration for the compiler.
+/// * `module` - The IRModule to compile.
+pub fn compile_module(config: CompilerConfig, module: IRModule) -> Result<RtModule, Error> {
```
Comment (hypercubestart) [inline rust/tvm/src/compiler/graph_rt.rs:89]: is there a test for this?

**Pull request merged**